legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/classpath/ClasspathUsingMavenFactory.java

```diff
@@ -551,39 +573,96 @@ private String findEngineVersion(File maven, File pom, File settingXml) throws I
+                    String cachedContent = Files.readString(legendLspCachedPom.toPath(), StandardCharsets.UTF_8);
+
+                    if (currentContent.equals(cachedContent))
+                    {
+                        this.server.logInfoToClient("Found cached " + id + " classpath, checking if still valid...");
+
+                        reloadClasspath = false;
+
+                        // can we still can read old files or do we depend on SNAPSHOTS... maybe jars where deleted
+                        for (String entry : classpath)
+                        {
+                            try
+                            {
+                                File file = new File(entry);
+                                if (!file.canRead() || entry.contains("-SNAPSHOT"))
+                                {
+                                    this.server.logInfoToClient("Need to reload " + id + " classpath.  Either an entry does not exist anymore or classpath contains -SNAPSHOT dependencies...");
+                                    reloadClasspath = true;
+                                    break;
+                                }
+                            }
+                            catch (Exception e)
+                            {
+                                reloadClasspath = true;
+                                break;
+                            }
+                        }
+                    }
+                }
             }
 
-            String classpath = Files.readString(legendLspClasspath.toPath(), StandardCharsets.UTF_8);
+            if (reloadClasspath)
+            {
+                this.server.logInfoToClient("Resolving " + id + " classpath invoking maven...");
+
+                Properties properties = new Properties();
+                properties.setProperty("mdep.outputFile", legendLspClasspath.getAbsolutePath());
+                properties.setProperty("mdep.pathSeparator", ";");
+
+                InvocationResult invocationResult = this.invokeMaven(maven, pom, settingXml, properties, "dependency:build-classpath");
+
+                if (invocationResult.getExitCode() != 0)
+                {
+                    String output = this.outputStream.toString(StandardCharsets.UTF_8);
+                    throw new RuntimeException("Unable to initialize Legend extensions.  Maven output:\n\n" + output);
+                }
+
+                Files.writeString(legendLspCachedPom.toPath(), currentContent, StandardCharsets.UTF_8);
+                classpath = Files.readString(legendLspClasspath.toPath(), StandardCharsets.UTF_8).split(";");
+            }
+            else
+            {
+                this.server.logInfoToClient("Reusing cached " + id + " classpath rather that invoking maven");
+            }
 
             LOGGER.info("{} classpath URLs used:  {}", id, classpath);
 
-            return Stream.of(classpath.split(";")).map(entry ->
+            return Stream.of(classpath).map(entry ->
```

Comment:
It could never be null, but to make it clearer, I merged/consolidated the behaviors around reload flag and the classpath array. If classpath is null,  classpath will be loaded thru maven.